File: ruby/bob/bob.rb
class Bob
  attr_reader :responder

  def initialize(grumpiness = NormalBobDay)
    @responder = grumpiness
  end

  def hey(words)
    responder.new(words).respond
  end
end

class NormalBobDay
  attr_reader :words

  def initialize(words)
    @words = words
  end

  def respond
    shouting || question || default
  end

  def shouting
    "Woah, chill out!" if shouting?
  end

  def shouting?
    words.upcase == words
  end

  def question
    "Sure." if question?
  end

  def question?
    words.end_with?("?")
  end

  def default
    "Whatever."
  end
end

File: javascript/bob/bob.js
var Bob = function(){
  'use strict';

  function question(input){
    return /\?$/.test(input);
  }

  function empty(input){
    return /^\s*$/.test(input);
  }

  function shouting(input){
    return (input.toUpperCase() === input) && (/[a-zA-Z]/.test(input))
  }

  this.hey = function(message) {
    if(shouting(message)){
      return "Woah, chill out!";
    } else if (question(message)) {
      return "Sure.";
    } else if (empty(message)) {
      return 'Fine. Be that way!';
    } else {
      return "Whatever.";
    }
  };
};
module.exports = Bob;
